### PR TITLE
fix: empty entries

### LINKED ISSUE #2876: v4.0.0-beta.0: Error: No dependency factory available for this dependency type: EntryDependency
- Operating System: Ubuntu
- Node Version: v14.15.0
- NPM Version: 6.14.8
- webpack Version: 5.9.0
- webpack-dev-server Version: v4.0.0-beta.0

- [x] This is a **bug**
- [ ] This is a **modification** request

### Code

```js
// webpack.config.js
  entry: {},
```

### Expected Behavior
No error

### Actual Behavior
Error:
```
<e> [webpack-dev-middleware] Error: No dependency factory available for this dependency type: EntryDependency
<e>     at Compilation.addModuleChain (.../node_modules/webpack/lib/Compilation.js:1632:5)
<e>     at Compilation._addEntryItem (.../node_modules/webpack/lib/Compilation.js:1742:8)
<e>     at Compilation.addEntry (.../node_modules/webpack/lib/Compilation.js:1673:8)
<e>     at .../node_modules/webpack-dev-server/lib/utils/DevServerPlugin.js:163:29
<e>     at new Promise (<anonymous>)
<e>     at .../node_modules/webpack-dev-server/lib/utils/DevServerPlugin.js:162:15
<e>     at Array.map (<anonymous>)
<e>     at .../node_modules/webpack-dev-server/lib/utils/DevServerPlugin.js:160:29
<e>     at Hook.eval [as callAsync] (eval at create (.../node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:38:17)
<e>     at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (.../node_modules/webpack/node_modules/tapable/lib/Hook.js:18:14)
```

### For Bugs; How can we reproduce the behavior?
I have only entrypoints generated by plugins (ModuleFederationPlugin) and I get this error [when I set empty object `{}` to `entry`](https://webpack.js.org/concepts/entry-points/#object-syntax).

As a workaround I can set path to empty js file. In this case, I get no error.

### PR DESCRIPTION
**Pull request by alexander-akait**
- [x] This is a **bugfix**
- [ ] This is a **feature**
- [ ] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

fixes #2876

### Motivation / Use-Case

Use better API

### Breaking Changes

No

### Additional Info

WIP on tests

### CODE REVIEW

**Comment by alexander-akait** [comment]
Rewriting tests will take a long time, we need to stop testing the internals of the webpack and focus on integration and e2e tests

**Comment by hiroppy** [comment]
ok so you will merge this pr even if CI fails?

**Comment by alexander-akait** [comment]
We need rewrite tests :disappointed:

File: lib/utils/DevServerPlugin.js [hunk 1]
```diff
@@ -153,24 +165,12 @@ class DevServerPlugin {
...
-                  {}, // global entry
-                  (err) => {
-                    if (err) return reject(err);
-                    resolve();
-                  }
-                );
-              })
-          )
-        )
-      );
```
Comment (ylemkimon) [inline lib/utils/DevServerPlugin.js:173]: The original intention was to allow modifying entries after they've been added.

Comment (alexander-akait) [reply lib/utils/DevServerPlugin.js:173]: Why? Can you clarify?

Comment (ylemkimon) [reply lib/utils/DevServerPlugin.js:173]: If we reference `additionalEntries` outside in the hook, they can be altered after hooks have added.

Comment (alexander-akait) [reply lib/utils/DevServerPlugin.js:173]: But why we need it? I can't undestand

**Comment by raix** [comment]
@alexander-akait when testing this branch on an app exposing a component I'm getting: 
image

I'll debug a bit, not sure if I have an issue in my dev server config after migrating to 4.x - the initial load of this app was fine, but it seems like it's not serving / exposing the component for other applications to consume.

**Comment by alexander-akait** [comment]
Not related to this issue, wrong public path

**Comment by raix** [comment]
Ok - I've migrated everything and should have correct configurations now - but I'll have a look at public path for clues,

Btw. would it make sense to create an example application app in the webpack-dev-server project or would that be a too complex test scenario?

The current test setup is two applications an app1 (consumer) and app2 (provider) using federated modules.
The main problem is in the provider application
Initial load:
image
On HMR:
image

**Comment by alexander-akait** [comment]
We have the issue for Module Federation, because it is not fully supported, please create reproducible test repo and put it there, this PR is not for helping, let's keep it only related to the problem

**Comment by raix** [comment]
Sorry, missed out on the issue

**Comment by csvan** [comment]
@alexander-akait anything we can do to help get this merged?

File: test/e2e/DevServer.test.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
+'use strict';
+
+const testBin = require('../helpers/test-bin');
+const isWebpack5 = require('../helpers/isWebpack5');
+
+describe('DevServer', () => {
+  const webpack5Test = isWebpack5 ? it : it.skip;
```
Comment (alexander-akait) [inline test/e2e/DevServer.test.js:7]: We should not skip some tests for webpack v4, because we still support it

Comment (snitin315) [reply test/e2e/DevServer.test.js:7]: Only skipping for webpack-5 specific tests. for ex when `target: [web, web-worker]` in config.

File: test/e2e/DevServer.test.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  it('should use client.path when custom', (done) => {
+    testBin('--config ./test/fixtures/dev-server/client-custom-path-config.js')
+      .then((output) => {
+        expect(output.exitCode).toEqual(0);
+        expect(output.stderr).toContain('&path=/custom/path');
+        done();
+      })
+      .catch(done);
+  });
+});
```
Comment (alexander-akait) [inline test/e2e/DevServer.test.js:92]: In future we should rewrite these tests on e2e with browser, i.e. run dev server, do some action and wait `HMR` or `reload` refresh using `puppeter` or `playwright`, we can do it in separate PR, should be interesting test

**Review by snitin315 — APPROVED** [review]
/cc @alexander-akait

**Comment by alexander-akait** [comment]
Now we need test it on IE 11 and ensure all works fine

**Pull request merged**